hv: do EPT mapping only for physical memory backed GPA on pre-launche…

ZideChen0 opened the pull request:
…d VMs

Currently for pre-launched VMs, HV intends to do EPT mapping for all GPA space,
which implies that it wastes HPA to back PCI hole and other vE820 entries that
are not of E820_TYPE_RAM type.

This patch only maps GPAs that are actually backed by physical memory to solve
this issue. Also, it fixes ve820 entries whose length is not aligned to 4K.

Tracked-On: #2587
Signed-off-by: Zide Chen <zide.chen@intel.com>
Acked-by: Anthony Xu <anthony.xu@intel.com>

dongyaozu commented on hypervisor/arch/x86/guest/vm.c:
```diff
@@ -130,6 +130,40 @@ static void create_prelaunched_vm_e820(struct acrn_vm *vm)
+ */
+static void create_ept_from_e820(struct acrn_vm *vm, const struct acrn_vm_config *vm_config)
+{
+	uint64_t base_hpa = vm_config->memory.start_hpa;
```
Can we reuse code of normal VM? Normal VM has similar operations to do, right?

dongyaozu replied:
Or the code SOS used. It also create EPT from E820.

ZideChen0 replied:
SOS maps all physical memory with UC attr and then go thru e820 and modify it to WB for E820_TYPE_RAM entries.  The duplicated code is not too much and it seems not easy to reuse them.

For UOS, EPT is created thru hypercall.

lifeix replied:
Normal world would add EPT mapping through hypercall one e820 by one. While SOS will create the EPT mapping for the whole address space to UC first and then modify the DM to WB according to e820. If we plan to only create EPT mapping for DM for SOS, we could reuse this code. The different is the hpa.

ZideChen0 replied:
> Normal world would add EPT mapping through hypercall one e820 by one. While SOS will create the EPT mapping for the whole address space to UC first and then modify the DM to WB according to e820. If we plan to only create EPT mapping for DM for SOS, we could reuse this code. The different is the hpa.

These few functions are small enough. If we want to reuse them seems we need to add a couple of "if () else()" which may make it more complicated.

BTW, seems it make sense to change the function name to make it consistent to SOS counterpart.
create_ept_from_e820() -> prepare_prelaunched_vm_memmap()

dongyaozu commented on hypervisor/arch/x86/guest/vm.c:
```diff
@@ -130,6 +130,40 @@ static void create_prelaunched_vm_e820(struct acrn_vm *vm)
+			uint64_t size = entry->length;
+
+			/* Merge memory under 1MB to lowmem so that it can be mapped with larger page size */
+			if (base_gpa == (uint64_t)MEM_1M) {
```
This assumes that the 0-1M memory has a separate e820 entry, and the next e820 entry doesn't contain 1M-2M space. This is strange?

ZideChen0 replied:
It only assumes there is an entry starts from 1M memory. It's for pre-launched VM whose ve820.c is statically defined in .c code.

The code could be complicated if make it generic and without assumption.

sainath14 replied:
MPTable use e820 reserved region so that OS does not claim and use that memory.
Instead of this approach, we can run thru all e820 entries and assign EPT/HPA for type E820_TYPE_RAM in this function.
Then in mptable_build API, we would add an EPT region for MPTable.
Drawbacks with this approach is that First 1 MB of GPA will be fragmented and would be using 4K EPT pages, and also we need to maintain where HPA has ended after every ept_mr_add somewhere in acrn_vm struct.
But code looks cleaner and sticks to the basics of E820 type definition.

ZideChen0 replied:
> MPTable use e820 reserved region so that OS does not claim and use that memory.
> Instead of this approach, we can run thru all e820 entries and assign EPT/HPA for type E820_TYPE_RAM in this function.
> Then in mptable_build API, we would add an EPT region for MPTable.
> Drawbacks with this approach is that First 1 MB of GPA will be fragmented and would be using 4K EPT pages, and also we need to maintain where HPA has ended after every ept_mr_add somewhere in acrn_vm struct.
> But code looks cleaner and sticks to the basics of E820 type definition.

Sainath,
Thanks for the comments. one more thing, we need to leave 1MB hole in E820, otherwise, the lowmem part EPT will be 4K EPT pages.

dongyaozu replied:
As discussed. the RAM section of e820 is better to start from 2MB rather than 1MB for simplicity... And simplify the code

dongyaozu commented:
@lifeix  can u comment too?

dongyaozu commented:
let us see if we can simplify with e820 contents simplification.

HuiHuangShi commented:
4 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:137<br> reason:Pointer param should be declared pointer to const. : vm<br><2> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:150<br> reason:Use of mixed mode arithmetic. :  (enum acpi_dmar_scope_type and unsigned char): entry -> type == 1U<br><3> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:160<br> reason:Casting operation to a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><4> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:160<br> reason:Casting operation on a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

ZideChen0 commented:
> let us see if we can simplify with e820 contents simplification.

There are only 5 e820 mem types (RAM, reserved, ACPI data, ACPI NVM, unusable) and it seems we have to keep the mptable e820 entry and stick to reserved type if we don't want to confuse people between mptable and ACPI.

If we don't want to merge 1MB memory to the low memory block, it will force it to 4KB EPT pages other than 2MB PDEs because either the HPA or GPA for the lowmem e820 entry is not 2MB aligned. We don't want this for performance concern, so we have to start the lowmem entry at GPA 2MB and leave a 1MB GPA hole. To further complicate it, because the length is (VM_CONFIG_MEM_SIZE  - 1M), then at the last 1MB, it creates PTE other than PDE. Optionally we can further hack it by reducing the entry size by another 1M.

It works, but the 1MB memory hole makes the e820 table look not normal.

        {       /* lowmem */
-               .baseaddr = 0x100000UL,  /* 1MB */
+               .baseaddr = 0x200000UL, /* 2MB */
-                .length   = 0x1FF00000UL, /* (VM_CONFIG_MEM_SIZE  - 1M) */
+               .length   = 0x1FE00000UL, /* (VM_CONFIG_MEM_SIZE  - 2M) */
                .type     = E820_TYPE_RAM
       }

And we still need to map the mptable e820 entry though it's of reserved type. So create_ept_from_e820() will be a little bit better but no big difference.

ZideChen0 commented:
My apologies. I just "push --force" to my branch and didn't realize that it updates the PR. I just reverted my changes, back to where we were reviewing.

HuiHuangShi commented:
102 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:part_config filename:/hypervisor/include/public/acrn_common.h func: offset:86<br> reason:Identifier reuse: component vs var. : size<br><2> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Parameter not declared on new line.          : param 2: vm_config<br><3> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Procedure definition has no associated prototype. : create_ept_from_e820<br><4> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:DU data flow anomalies found.<br><5> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Referenced parameter has possible clear path. : (vm)->arch_vm.nworld_eptp<br><6> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Referenced parameter has possible clear path. : (vm)->arch_vm<br><7> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Referenced parameter has possible clear path. : (vm)->e820_entries<br><8> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:No prefix for global function.<br><9> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Procedure is not structured.<br><10> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:* or & not attached to type in declaration.<br><11> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Procedure contains essential knots.<br><12> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Pointer param should be declared pointer to const. : vm<br><13> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:No space between * or & and name in declaration.<br><14> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Identifier reuse: proc param vs proc param. : vm_config<br><15> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Identifier reuse: proc param vs proc param. : vm<br><16> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Pointer param should be declared const pointer. : vm<br><17> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:138<br> reason:Pointer param should be declared const pointer. : vm_config<br><18> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:140<br> reason:DU anomaly, variable value is not used. : base_hpa<br><19> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:140<br> reason:Name redeclared in another C name space (MR). : base_hpa<br><20> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:140<br> reason:{ ... } contents not indented by 2 spaces.<br><21> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:140<br> reason:No comment after declaration.<br><22> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:140<br> reason:Tab character in source.<br><23> config:sharing_config filename:/hypervisor/arch/x86/guest/vm.c func:get_vm_bsp_pcpu_id offset:140<br> reason:Tab character in source.<br><24> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:141<br> reason:Name found with length less than 2.          : i<br><25> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:141<br> reason:Identifier reuse: variable vs variable. : i<br><26> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:141<br> reason:No comment after declaration.<br><27> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:141<br> reason:Local not initialised at declaration. : i<br><28> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:143<br> reason:More than one exit from a loop.<br><29> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:143<br> reason:{ or } not on line by itself.<br><30> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:143<br> reason:{ ... } contents not indented by 2 spaces.<br><31> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:143<br> reason:Use of numeric literal in expression.        : 0U<br><32> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:144<br> reason:No space between * or & and name in declaration.<br><33> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:144<br> reason:No comment after declaration.<br><34> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:create_ept_from_e820 offset:144<br> reason:* or & not attached to type in declaration.<br>.....<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

ZideChen0 commented:
I just updated the PR as requested, and also changed function name to make it  consistent to similar function for SOS: create_ept_from_e820() -> prepare_prelaunched_vm_memmap()

The E820 on MRB board looks like this.

BIOS-e820: [mem 0x0000000000000000-0x00000000000effff] usable
BIOS-e820: [mem 0x00000000000f0000-0x00000000000fffff] reserved
BIOS-e820: [mem 0x0000000000200000-0x00000000200fffff] usable
BIOS-e820: [mem 0x0000000020100000-0x00000000bfffffff] reserved
BIOS-e820: [mem 0x00000000e0000000-0x00000000ffffffff] reserved

HuiHuangShi commented:
7 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:138<br> reason:Pointer param should be declared pointer to const. : vm<br><2> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:Use of mixed mode arithmetic. :  (enum acpi_dmar_scope_type and unsigned char): entry -> type == 1U<br><3> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:152<br> reason:Casting operation to a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><4> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:152<br> reason:Casting operation on a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><5> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:159<br> reason:Use of mixed mode arithmetic. :  (enum acpi_dmar_scope_type and unsigned char): entry -> type != 1U<br><6> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:160<br> reason:Casting operation to a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><7> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:160<br> reason:Casting operation on a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

ZideChen0 commented:
> ointer param should be declared pointer to const. : vm

All are "false positive" or "to Deviate"

HuiHuangShi commented:
out of time

dongyaozu commented on hypervisor/arch/x86/configs/apl-mrb/ve820.c:
```diff
@@ -7,33 +7,33 @@
-		.length   = 0x1FF00000UL,
+	{	/* lowmem */
+		.baseaddr = 0x200000UL,		/* 2MB */
+		.length   = 0x1FF00000UL,	/* 511MB */
```
Then here it should be 512MB, not?

ZideChen0 replied:
Total GPA is 512M, the first 0-1MB is in the first two entries. So here is 511MB.

dongyaozu commented on hypervisor/arch/x86/configs/apl-mrb/ve820.c:
```diff
@@ -7,33 +7,33 @@
-		.length   = 0xa0000000UL,
+	{	/* between lowmem and PCI hole */
+		.baseaddr = 0x20100000UL,	/* 513MB */
+		.length   = 0x9FF00000UL,	/* 2559GB */
```
should be 256GB, not?

ZideChen0 replied:
Because of the 1MB memory hole, the lowmem entry is 2MB-- 0x20100000 (512MB + 1MB). If we keep this entry as 2560MB, it will eat 1MB into the PCI hole (0xc0000000). So I cut 1MB from this reserved entry and keep PCI hole clean.

dongyaozu commented on hypervisor/arch/x86/configs/apl-mrb/ve820.c:
```diff
@@ -7,33 +7,33 @@
-		.baseaddr = 0x20000000UL,
-		.length   = 0xa0000000UL,
+	{	/* between lowmem and PCI hole */
+		.baseaddr = 0x20100000UL,	/* 513MB */
```
why it starts from 513MB? should be 512MB.

ZideChen0 replied:
Because of the 1MB memory hole. So the 512MB memory is spread out: 0-1MB, 2MB-513MB. 
This looks weird. 

The alternative is to cut 1MB from the usable RAM and let it finish at 512MB. Then it makes the total available guest memory to be 511MB, still looks not normal.

Or we can keep the original solution to merge 1MB into the lowmem entry.

dongyaozu replied:
No, it should be 0-960KB, 2MB-512MB.

dongyaozu replied:
All the segment start and end address should be 2MB aligned, except the 0-1MB space

ZideChen0 replied:
OK, I can update accordingly.

dongyaozu commented on hypervisor/arch/x86/configs/apl-mrb/ve820.c:
```diff
@@ -7,33 +7,33 @@
-		.length   = 0x20000000UL,
+	{	/* between PCI hole and 4GB */
+		.baseaddr = 0xe0000000UL,	/* 3.5GB */
+		.length   = 0x20000000UL,	/* 512MB */
```
This is good

dongyaozu replied:
BTW, should we use (baseaddr, length), or (start_address, end_address)?

ZideChen0 replied:
There are two reasons for (base, size, type): 1) The Linux zero page structure is defined in this way so we can easily copy our e820 to zero page. 2) "Address Range Descriptor" in INT 15h E820h has same content.

HuiHuangShi commented:
111 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func: offset:145<br> reason:Less than 2 blank lines separating procs.<br><2> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Parameter not declared on new line.          : param 2: vm_config<br><3> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Procedure definition has no associated prototype. : prepare_prelaunched_vm_memmap<br><4> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:DU data flow anomalies found.<br><5> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:No space between * or & and name in declaration.<br><6> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Identifier reuse: proc param vs proc param. : vm_config<br><7> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Identifier reuse: proc param vs proc param. : vm<br><8> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:No prefix for global function.<br><9> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Procedure is not structured.<br><10> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:* or & not attached to type in declaration.<br><11> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Procedure contains essential knots.<br><12> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Pointer param should be declared pointer to const. : vm<br><13> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Function starts with lower case character.   : prepare_prelaunched_vm_memmap<br><14> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Pointer param should be declared const pointer. : vm_config<br><15> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Pointer param should be declared const pointer. : vm<br><16> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Referenced parameter has possible clear path. : (vm)->e820_entries<br><17> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Referenced parameter has possible clear path. : (vm)->arch_vm.nworld_eptp<br><18> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:150<br> reason:DU anomaly, variable value is not used. : base_hpa<br><19> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:150<br> reason:No comment after declaration.<br><20> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:150<br> reason:{ ... } contents not indented by 2 spaces.<br><21> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:150<br> reason:Name redeclared in another C name space (MR). : base_hpa<br><22> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:150<br> reason:Tab character in source.<br><23> config:sharing_config filename:/hypervisor/arch/x86/guest/vm.c func:get_vm_bsp_pcpu_id offset:150<br> reason:Tab character in source.<br><24> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:Name found with length less than 2.          : i<br><25> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:Identifier reuse: variable vs variable. : i<br><26> config:sharing_config filename:/hypervisor/arch/x86/guest/vm.c func:get_vm_bsp_pcpu_id offset:151<br> reason:Tab character in source.<br><27> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:Tab character in source.<br><28> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:No comment after declaration.<br><29> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:Local not initialised at declaration. : i<br><30> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:153<br> reason:Loop termination not a simple constant test.<br><31> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:153<br> reason:{ or } not on line by itself.<br><32> config:sharing_config filename:/hypervisor/arch/x86/guest/vm.c func:get_vm_bsp_pcpu_id offset:153<br> reason:Tab character in source.<br><33> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:153<br> reason:Tab character in source.<br><34> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:153<br> reason:Use of numeric literal in expression.        : 0U<br>.....<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

HuiHuangShi commented:
111 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func: offset:145<br> reason:Less than 2 blank lines separating procs.<br><2> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Parameter not declared on new line.          : param 2: vm_config<br><3> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Procedure definition has no associated prototype. : prepare_prelaunched_vm_memmap<br><4> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:DU data flow anomalies found.<br><5> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:No space between * or & and name in declaration.<br><6> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Identifier reuse: proc param vs proc param. : vm_config<br><7> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Identifier reuse: proc param vs proc param. : vm<br><8> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:No prefix for global function.<br><9> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Procedure is not structured.<br><10> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:* or & not attached to type in declaration.<br><11> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Procedure contains essential knots.<br><12> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Pointer param should be declared pointer to const. : vm<br><13> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Function starts with lower case character.   : prepare_prelaunched_vm_memmap<br><14> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Pointer param should be declared const pointer. : vm_config<br><15> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Pointer param should be declared const pointer. : vm<br><16> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Referenced parameter has possible clear path. : (vm)->e820_entries<br><17> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:148<br> reason:Referenced parameter has possible clear path. : (vm)->arch_vm.nworld_eptp<br><18> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:150<br> reason:DU anomaly, variable value is not used. : base_hpa<br><19> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:150<br> reason:No comment after declaration.<br><20> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:150<br> reason:{ ... } contents not indented by 2 spaces.<br><21> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:150<br> reason:Name redeclared in another C name space (MR). : base_hpa<br><22> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:150<br> reason:Tab character in source.<br><23> config:sharing_config filename:/hypervisor/arch/x86/guest/vm.c func:get_vm_bsp_pcpu_id offset:150<br> reason:Tab character in source.<br><24> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:Name found with length less than 2.          : i<br><25> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:Identifier reuse: variable vs variable. : i<br><26> config:sharing_config filename:/hypervisor/arch/x86/guest/vm.c func:get_vm_bsp_pcpu_id offset:151<br> reason:Tab character in source.<br><27> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:Tab character in source.<br><28> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:No comment after declaration.<br><29> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:151<br> reason:Local not initialised at declaration. : i<br><30> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:153<br> reason:Loop termination not a simple constant test.<br><31> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:153<br> reason:{ or } not on line by itself.<br><32> config:sharing_config filename:/hypervisor/arch/x86/guest/vm.c func:get_vm_bsp_pcpu_id offset:153<br> reason:Tab character in source.<br><33> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:153<br> reason:Tab character in source.<br><34> config:part_config filename:/hypervisor/arch/x86/guest/vm.c func:prepare_prelaunched_vm_memmap offset:153<br> reason:Use of numeric literal in expression.        : 0U<br>.....<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>
